pydantic_ai_slim/pydantic_ai/output.py

```diff
@@ -0,0 +1,289 @@
+
+from collections.abc import Awaitable, Sequence
+from dataclasses import dataclass
+from typing import Any, Callable, Generic, Literal, Union
+
+from typing_extensions import TypeAliasType, TypeVar, get_args, get_origin
+from typing_inspection import typing_objects
+from typing_inspection.introspection import is_union_origin
+
+from .messages import RetryPromptPart
+from .tools import RunContext
+
+OutputDataT = TypeVar('OutputDataT', default=str, covariant=True)
+"""Covariant type variable for the result data type of a run."""
+
+T = TypeVar('T')
+
+OutputMode = Literal['text', 'tool', 'model_structured', 'prompted_structured', 'tool_or_text']
+"""All output modes."""
+StructuredOutputMode = Literal['tool', 'model_structured', 'prompted_structured']
+"""Output modes that can be used for structured output. Used by ModelProfile.default_structured_output_mode"""
+
+
+class ToolRetryError(Exception):
+    """Internal exception used to signal a `ToolRetry` message should be returned to the LLM."""
+
+    def __init__(self, tool_retry: RetryPromptPart):
+        self.tool_retry = tool_retry
+        super().__init__()
+
+
+@dataclass(init=False)
+class ToolOutput(Generic[OutputDataT]):
+    """Marker class to use a tool for output and optionally customize the tool.
+
+    Example:
+    ```python
+    from pydantic import BaseModel
+
+    from pydantic_ai import Agent, ToolOutput
+
+
+    class Fruit(BaseModel):
+        name: str
+        color: str
+
+
+    class Vehicle(BaseModel):
+        name: str
+        wheels: int
+
+
+    agent = Agent(
+        'openai:gpt-4o',
+        output_type=[
+            ToolOutput(Fruit, name='return_fruit'),
+            ToolOutput(Vehicle, name='return_vehicle'),
+        ],
+    )
+    result = agent.run_sync('What is a banana?')
+    print(repr(result.output))
+    #> Fruit(name='banana', color='yellow')
+    ```
+    """
+
+    output: OutputTypeOrFunction[OutputDataT]
+    name: str | None
+    description: str | None
+    max_retries: int | None
+    strict: bool | None
+
+    def __init__(
+        self,
+        type_: OutputTypeOrFunction[OutputDataT],
+        *,
+        name: str | None = None,
+        description: str | None = None,
+        max_retries: int | None = None,
+        strict: bool | None = None,
+    ):
+        self.output = type_
+        self.name = name
+        self.description = description
+        self.max_retries = max_retries
+        self.strict = strict
+
+
+@dataclass(init=False)
+class ModelStructuredOutput(Generic[OutputDataT]):
+    """Marker class to use the model's built-in structured outputs functionality for outputs and optionally customize the name and description.
+
+    Example:
+    ```python
+    from pydantic import BaseModel
+
+    from pydantic_ai import Agent, ModelStructuredOutput
+
+
+    class Fruit(BaseModel):
+        name: str
+        color: str
+
+
+    class Vehicle(BaseModel):
+        name: str
+        wheels: int
+
+
```

Comment:
again, import rather than duplicating.

Reply:
This is the docstring though, so I'm not sure where you're thinking of putting them and importing them from. I think the docstrings should be complete without having to refer to other places (unless those other places have their own docs).